vortex-layout/src/layouts/view/reader/mod.rs

```diff
@@ -0,0 +1,211 @@
+//  SPDX-License-Identifier: Apache-2.0
+//  SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+mod array;
+mod filter;
+mod pruning;
+
+use std::collections::BTreeSet;
+use std::ops::Range;
+use std::sync::{Arc, OnceLock};
+
+use futures::FutureExt;
+use futures::future::{BoxFuture, Shared};
+use vortex_array::ArrayContext;
+use vortex_array::arrays::BinaryView;
+use vortex_array::stats::Precision;
+use vortex_buffer::{Buffer, ByteBuffer};
+use vortex_dtype::{DType, FieldMask};
+use vortex_error::{SharedVortexResult, VortexResult};
+use vortex_expr::{ExprRef, LikeVTable, ScopeDType};
+
+use crate::layouts::view::ViewLayout;
+use crate::layouts::view::reader::array::ViewProjection;
+use crate::layouts::view::reader::filter::ViewFilter;
+use crate::segments::{SegmentId, SegmentSource};
+use crate::{
+    ArrayEvaluation, LayoutReader, MaskEvaluation, NoOpPruningEvaluation, PruningEvaluation,
+};
+
+type BinaryViewFuture = Shared<BoxFuture<'static, SharedVortexResult<Buffer<BinaryView>>>>;
+type BufferFuture = Shared<BoxFuture<'static, SharedVortexResult<ByteBuffer>>>;
+
+/// Handle for fetching a buffer by index.
+#[derive(Clone)]
+pub(crate) struct FetchBuffers {
+    name: Arc<str>,
+    segment_source: Arc<dyn SegmentSource>,
+    buffer_ids: Arc<[SegmentId]>,
+    buffers: Vec<OnceLock<BufferFuture>>,
```

Comment:
I _think_ (given how horrible the implicit prefetching is...) you might want to just create all these futures up-front.

Creation causes registration for prefetching.

Only when you first _poll_ them does the segment get marked as required.

By deferring here with a segment source, you're precluding these buffers from ever being prefetched.

Reply:
Yea I was thinking about that, but I thought part of the goal here is that if we can prune using the views some of the buffers may not need to be fetched _at all_. Maybe it's not worth it to lose the prefetching though

Reply:
These are the kind of behaviour bugs I really hate... 

Luckily, Rust lets us solve them at compile time!

We should pass in &dyn SegmentSource, because the resulting evaluation needs to have a static lifetime, it would be impossible for someone to delay getting a segment future.

Reply:
Here's an example of what I mean. Not sure if it's the right thing to do though... https://github.com/vortex-data/vortex/pull/3858